src/stp/new_tactics/ChipAtPos.cpp

```diff
@@ -0,0 +1,107 @@
+
+namespace rtt::ai::stp::tactic {
+
+ChipAtPos::ChipAtPos() {
+    // Create state machine of skills and initialize first skill
+    skills = rtt::collections::state_machine<Skill, Status, StpInfo>{skill::Rotate(), skill::Chip()};
+    skills.initialize();
+}
+
+void ChipAtPos::onInitialize() noexcept {}
+
+void ChipAtPos::onUpdate(Status const &status) noexcept {}
+
+void ChipAtPos::onTerminate() noexcept {
+    // Call terminate on all skills
+    for (auto &x : skills) {
+        x->terminate();
+    }
+}
+
+StpInfo ChipAtPos::calculateInfoForSkill(StpInfo const &info) noexcept {
+    if (!info.getBall() || !info.getRobot() || !info.getField()) {
+        RTT_WARNING("No Ball, Robot or Field present in StpInfo");
+        return {};
+    }
+
+    StpInfo skillStpInfo = info;
+
+    // Calculate the angle the robot needs to aim
+    double angleToBall = (info.getPosition().second - info.getBall()->get()->getPos()).angle();
+    skillStpInfo.setAngle(angleToBall);
+
+    // Calculate the distance and the chip force
+    double distanceBallToTarget = (info.getBall()->get()->getPos() - info.getPosition().second).length();
+    skillStpInfo.setKickChipVelocity(determineChipForce(distanceBallToTarget, skillStpInfo.getKickChipType()));
+
+    // When chipping, we don't need to activate the dribbler
+    skillStpInfo.setDribblerSpeed(0);
+
+    return skillStpInfo;
+}
+
+/// Determine how fast we should chip for a pass at a given distance
+double ChipAtPos::determineChipForce(double distance, KickChipType desiredBallSpeedType) noexcept {
+    const double maxPowerDist = rtt::ai::Constants::MAX_POWER_KICK_DISTANCE();
+
+    double velocity = 0;
+    switch (desiredBallSpeedType) {
+        case DRIBBLE_KICK: {
+            velocity = sqrt(distance) * rtt::ai::Constants::MAX_KICK_POWER() / (sqrt(maxPowerDist) * 1.5);
+            break;
+        }
+        case BALL_PLACEMENT: {
+            if (distance > 2.5) {
+                velocity = Constants::GRSIM() ? 6.01 : 2.01;
+            } else {
+                velocity = Constants::GRSIM() ? 3.01 : 1.01;
+            }
+            break;
+        }
+        case PASS: {
+            if (distance >= maxPowerDist) {
+                velocity = Constants::MAX_KICK_POWER();
+            } else if (Constants::GRSIM()) {
+                velocity = std::min(1.4 * distance / maxPowerDist * Constants::MAX_KICK_POWER(), Constants::DEFAULT_KICK_POWER());
+            } else {
+                velocity = std::min(distance / maxPowerDist * Constants::MAX_KICK_POWER(), Constants::DEFAULT_KICK_POWER() * 0.7);
+            }
+            break;
+        }
+        case MAX_SPEED: {
+            velocity = rtt::ai::Constants::MAX_KICK_POWER();
+            break;
+        }
+        default: {
+            velocity = rtt::ai::Constants::MAX_KICK_POWER();
+        }
+    }
+
+    // limit the output to the max chip speed
+    return std::min(std::max(velocity, 1.01), rtt::ai::Constants::MAX_KICK_POWER());
+}
```

Comment:
Warning: magic numbers detected. Magic is forbidden in our code. Also, can you write some documentation of comments why grsim is special? As I can see, you just divide the speed by 2, so maybe it's better to do that in the cases.

Reply:
We definitely should refactor this for both kicking and chipping